File: project/security/src/main/java/org/tby/fourdk/security/Role.java
package org.tby.fourdk.security;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class Role {

    private String name;

    private List<Permission> associatedPermissions;

    public Role(String role) {
        this.name = role;
    }

    public Role(String name, List<Permission> associatedPermissions) {
        this.associatedPermissions = associatedPermissions;
        this.name = name;
    }

    public boolean isThereSomeUnknownPermissions(List<Permission> permissions) {
        return !permissions.containsAll(this.associatedPermissions);
    }

    public UUID getId() {
        return UUID.nameUUIDFromBytes(this.name.getBytes());
    }


    public String getName() {
        return name;
    }

    public List<Permission> getAssociatedPermissions() {
        return associatedPermissions != null ? associatedPermissions : new ArrayList<>();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Role role = (Role) o;
        return Objects.equals(name, role.name) && Objects.equals(associatedPermissions, role.associatedPermissions);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, associatedPermissions);
    }

    @Override
    public String toString() {
        return "Role{" +
                "name='" + name + '\'' +
                ", associatedPermissions=" + associatedPermissions +
                '}';
    }
}

File: rust/project/security-rust/src/dddk/security/role.rs
use crate::dddk::security::permission::Permission;

#[derive(Eq, Hash)]
pub struct Role {
    name: String,
    associated_permissions: Vec<Permission>,
}

impl Role {
    pub fn new(name: String, associated_permissions: Vec<Permission>) -> Role {
        Role {
            name,
            associated_permissions,
        }
    }

    pub fn get_name(&self) -> &String {
        &self.name
    }

    pub fn get_associated_permissions(&self) -> &Vec<Permission> {
        &self.associated_permissions
    }
}

impl Clone for Role {
    fn clone(&self) -> Self {
        let name = self.get_name().clone();
        let associated_permissions = self.get_associated_permissions().clone();
        Role::new(name, associated_permissions)
    }
}

impl PartialEq<Self> for Role {
    fn eq(&self, other: &Self) -> bool {
        self.get_name() == other.get_name() &&
            self.associated_permissions == other.associated_permissions
    }
}
